rfcs/000-caller-side-managed-buffers.md

```diff
@@ -0,0 +1,833 @@
+    let status = conn.handle_tls_records(&mut incoming_tls, &mut outgoing_tls)?;
+
+    match status.state {
+        State::MustTransmitTlsData => {
+            if status.caps.may_encrypt_early_data {
+                // application-specific logic goes here; contrived client-like example:
+
+                let request = prepare_request();
+                conn.encrypt_early_data(&request, &mut outgoing_tls);
+            }
+
+            if status.caps.may_encrypt_app_data {
+                // opportunistically send application data during the handshake
+                // application-specific logic goes here; contrived client-like example:
+
+                let request = prepare_request();
+                conn.encrypt_app_data(&request, &mut outgoing_tls);
+            }
+
+            socket.write_all(&outgoing_tls)?;
+            outgoing_tls.clear();
+        },
+
+        State::NeedsMoreTlsData => {
+            let read = socket.read(incoming_tls.unfilled())?;
+            incoming_tls.advance(read);
+        },
+
+        State::ReceivedAppData => {
+            // application-specific logic goes here; contrived server-like example:
+
+            let request = conn.decrypt_incoming(&mut incoming_tls).next().unwrap()?;
+            handle(request);
+        },
+
+        State::ReceivedEarlyData => {
+            // application-specific logic goes here; contrived server-like example:
+
+            let request = conn.decrypt_early_data(&mut incoming_tls).unwrap()?;
+            handle(request);
+        },
+
+        State::TrafficTransit => {
+            // handshake complete; break out of the handshake loop
+            break
+        },
+    }
+}
+
+// post-handshake application logic goes here
+```
+
+## Resizing buffers
+
+One of the goals of this new API is to support resizing of all the involved buffers.
+An example is provided below:
+
+```rust
+let handshake_size: usize = 64 * 1024;
+let mut outgoing_tls = Vec::with_capacity(handshake_size);
+let mut incoming_tls = IncomingTls::new(vec![0; handshake_size]);
+
+// complete the handshake process
+loop {
+    // ..
+}
+
+let post_handshake_size: usize = 4096;
+if incoming_tls.filled().is_empty() {
+    let mut buf = incoming_tls.into_inner();
+    buf.truncate(post_handshake_size);
+    incoming_tls = IncomingTls::new(buf);
+}
+
+if outgoing_tls.len() <= post_handshake_size {
+    outgoing_tls.truncate(post_handshake_size);
+}
+
+// ..
+```
+
+In the case of `IncomingTls`, care must be taken to check that the buffer is empty prior to resizing.
+
+In the case of `outgoing_tls`, the buffer can be resized while it contains data but
+care must be taken not to discard existing data if truncating it.
+
+# Implementation details
+
+This section contains implementation details about the new API.
+
+## Partially discarding app data records
+
+To implement an equivalent of `rustls::Stream` API using the `LlConnection` types and, in particular, to implement the `io::Read` trait,
```

Comment:
Let's place an equivalent of `rustls::Stream` out of scope for now?

Reply:
to clarify this is not proposing neither adding a new `Stream` on top of `LlConnectionCommon` nor modifying the existing implementation. it was more along the lines of me attempting to implement `Stream` on top of `LlConnectionCommon` to see if the API worked well enough to implement `Stream`.

I now think it's not absolutely necessary to support partially discarding records at the `LlConnection` layer to implement `Stream`. any app-data record fragment that doesn't fit in `Read::read`'s buffer could be buffered into `Stream` until the next `Read::read`'s call. as `Stream` is more of a convenience API and not a "max performance" API I think that may be an ok trade-off to keep the `LlConnection` API simpler.